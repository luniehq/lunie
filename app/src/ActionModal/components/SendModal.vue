<template>
  <ActionModal
    id="send-modal"
    ref="actionModal"
    :validate="validateForm"
    :amount="amount"
    title="Send"
    submission-error-prefix="Sending tokens failed"
    :transaction-type="messageType.SEND"
    :transaction-data="transactionData"
    :selected-denom="selectedToken"
    :notify-message="notifyMessage"
    feature-flag="send"
    :disabled="
      $asyncComputed.transactionData.updating ||
      $asyncComputed.transactionData.error
    "
    @close="clear"
    @txIncluded="onSuccess"
  >
    <TmFormGroup
      :error="$v.address.$error && $v.address.$invalid"
      class="action-modal-form-group"
      field-id="send-address"
      field-label="Send To"
    >
      <TmField
        id="send-address"
        ref="sendAddress"
        v-model="address"
        v-focus
        type="text"
        placeholder="Address or Starname"
        @change.native="trimSendAddress"
        @keyup.enter.native="refocusOnAmount"
      />
      <TmFormMsg
        v-if="$v.address.$error && !$v.address.required"
        name="Address"
        type="required"
      />
      <TmFormMsg
        v-else-if="$v.address.$error && !$v.address.validAddress"
        name="Address"
        type="custom"
        msg="doesn't have a format known by Lunie"
      />
      <span v-if="$asyncComputed.transactionData.updating" class="memo-span">
        Getting address for Starname...
      </span>
      <TmFormMsg
        v-else-if="$asyncComputed.transactionData.error"
        msg="could not be resolved to an address"
        name="Starname"
        type="custom"
      />
      <span v-else class="memo-span">
        Lunie supports
        <a href="https://starname.me/" target="_blank">IOV Starnames</a>
      </span>
    </TmFormGroup>
    <TmFormGroup
      id="form-group-amount"
      :error="$v.amount.$error && $v.amount.$invalid"
      class="action-modal-form-group"
      field-id="amount"
      field-label="Amount"
    >
      <TmFieldGroup>
        <TmField
          id="amount"
          ref="amount"
          v-model="amount"
          class="tm-field-addon"
          placeholder="0"
          type="number"
          @keyup.enter.native="enterPressed"
        />
        <TmField
          id="token"
          v-model="selectedToken"
          :title="`Select the token you wish to operate with`"
          :options="getDenoms"
          class="tm-field-token-selector"
          placeholder="Select the token"
          type="select"
        />
        <TmBtn
          type="button"
          class="addon-max"
          value="Set Max"
          @click.native="setMaxAmount()"
        />
      </TmFieldGroup>
      <TmFormMsg
        v-if="selectedBalance.amount === 0"
        :msg="`doesn't have any ${selectedToken}s`"
        name="Wallet"
        type="custom"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && (!$v.amount.required || amount === 0)"
        name="Amount"
        type="required"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && !$v.amount.decimal"
        name="Amount"
        type="numeric"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && !$v.amount.max"
        type="custom"
        :msg="`You don't have enough ${selectedToken}s to proceed.`"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && !$v.amount.min"
        :min="smallestAmount"
        name="Amount"
        type="min"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && !$v.amount.maxDecimals"
        name="Amount"
        type="maxDecimals"
      />
      <TmFormMsg
        v-else-if="isMaxAmount()"
        msg="You are about to use all your tokens for this transaction. Consider leaving a little bit left over to cover the network fees."
        type="custom"
        class="tm-form-msg--desc max-message"
      />
    </TmFormGroup>
    <TmFormGroup
      id="memo"
      :error="$v.memo.$error && $v.memo.$invalid"
      class="action-modal-group"
      field-id="memo"
      field-label="Memo"
    >
      <TmField
        id="memo"
        v-model="memo"
        type="text"
        @keyup.enter.native="enterPressed"
      />
      <span class="memo-span"
        >To learn more about how to use the memo field, read
        <a
          href=" https://intercom.help/lunie/en/articles/3776563-using-the-memo-option-when-sending-tokens-to-and-from-exchanges"
          rel="noopener norefferer"
          target="_blank"
          >our guide</a
        >.</span
      >
      <TmFormMsg
        v-if="$v.memo.$error && !$v.memo.maxLength"
        name="Memo"
        type="maxLength"
        :max="max_memo_characters"
      />
    </TmFormGroup>
  </ActionModal>
</template>

<script>
import gql from "graphql-tag"
import b32 from "scripts/b32"
import { required, decimal, maxLength } from "vuelidate/lib/validators"
import { SMALLEST } from "src/scripts/num"
import { mapGetters } from "vuex"
import TmFormGroup from "src/components/common/TmFormGroup"
import TmField from "src/components/common/TmField"
import TmFieldGroup from "src/components/common/TmFieldGroup"
import TmBtn from "src/components/common/TmBtn"
import TmFormMsg from "src/components/common/TmFormMsg"
import ActionModal from "./ActionModal"
import { messageType } from "../../components/transactions/messageTypes"
import config from "src/../config"
import { UserTransactionAdded } from "src/gql"
import BigNumber from "bignumber.js"
import { formatAddress } from "src/filters"
import { isStarname, resolveStarname } from "src/../../common/starname"

const defaultMemo = ""

const isPolkadotAddress = (address) => {
  const polkadotRegexp = /^(([0-9a-zA-Z]{47})|([0-9a-zA-Z]{48}))$/
  return polkadotRegexp.test(address)
}

export default {
  name: `send-modal`,
  components: {
    TmField,
    TmFieldGroup,
    TmFormGroup,
    TmFormMsg,
    ActionModal,
    TmBtn,
  },
  props: {
    denoms: {
      type: Array,
      required: true,
    },
  },
  data: () => ({
    address: ``,
    amount: config.development ? 0.000001 : null, // dev life, hard life > make simple
    memo: defaultMemo,
    max_memo_characters: 256,
    isFirstLoad: true,
    selectedToken: undefined,
    balances: [],
    messageType,
    smallestAmount: SMALLEST,
    networkFeesLoaded: false,
  }),
  computed: {
    ...mapGetters([`network`, `networks`, `stakingDenom`]),
    ...mapGetters({ userAddress: `address` }),
    selectedBalance() {
      return (
        this.balances.find(({ denom }) => denom === this.selectedToken) || {
          amount: 0,
        }
      )
    },
    notifyMessage() {
      return {
        title: `Successful Send`,
        body: `Successfully sent ${this.amount} ${
          this.selectedToken
        }s to ${formatAddress(this.address)}`,
      }
    },
    getDenoms() {
      return this.denoms
        ? this.denoms.map((denom) => (denom = { key: denom, value: denom }))
        : []
    },
    // TODO: maxAmount should be handled from ActionModal
    maxAmount() {
      if (this.networkFeesLoaded) {
        return this.maxDecimals(
          this.selectedBalance.amount - this.networkFees.transactionFee.amount,
          6
        )
      } else {
        return this.maxDecimals(this.selectedBalance.amount, 6)
      }
    },
  },
  asyncComputed: {
    async transactionData() {
      if (isNaN(this.amount) || !this.address || !this.selectedToken) {
        return {}
      }
      return {
        type: messageType.SEND,
        to: [this.address],
        from: [this.userAddress],
        amount: {
          amount: this.amount,
          denom: this.selectedToken,
        },
        memo: this.memo,
      }
    },
  },
  watch: {
    // we set the amount in the input to zero every time the user selects another token so they
    // realize they are dealing with a different balance each time
    selectedToken: function () {
      if (!this.isFirstLoad) {
        this.amount = 0
      } else {
        this.isFirstLoad = false
      }
    },
    balances: function (balances) {
      // if there is already a token selected don't reset it
      if (this.selectedToken) return

      // in case the account has no balances we will display the staking denom received from the denom query
      if (balances.length === 0) {
        this.selectedToken = this.stakingDenom
      } else {
        this.selectedToken = balances[0].denom
      }
    },
    address: async function (address) {
      if (isStarname(this.address)) {
        // resolve starname to address
        const recipientAddress = await resolveStarname(
          this.address,
          this.network
        )
        this.setAddress(recipientAddress)
      }
    },
  },
  mounted() {
    this.$apollo.queries.balances.refetch()
  },
  methods: {
    open(denom = undefined) {
      if (denom) {
        this.selectedToken = denom
      } else {
        this.selectedToken =
          this.balances && this.balances.length > 0
            ? this.balances[0].denom
            : this.stakingDenom
      }
      this.$refs.actionModal.open()
    },
    onSuccess(event) {
      this.$emit(`success`, event)
    },
    validateForm() {
      this.$v.$touch()

      return !this.$v.$invalid
    },
    clear() {
      this.$v.$reset()

      this.address = undefined
      this.amount = undefined
      this.memo = defaultMemo
      this.sending = false
    },
    setMaxAmount() {
      this.amount = this.maxAmount
    },
    setAddress(recipientAddress) {
      this.address = recipientAddress
    },
    isMaxAmount() {
      if (this.selectedBalance.amount === 0) {
        return false
      } else {
        return parseFloat(this.amount) === this.maxAmount
      }
    },
    token() {
      if (!this.selectedToken) return ``

      return this.selectedToken
    },
    bech32Validate(param) {
      try {
        b32.decode(param)
        return true
      } catch (error) {
        return false
      }
    },
    enterPressed() {
      this.$refs.actionModal.validateChangeStep()
    },
    trimSendAddress() {
      this.address = this.$refs.sendAddress.value.trim()
    },
    refocusOnAmount() {
      this.$refs.amount.$el.focus()
    },
    maxDecimals(value, decimals) {
      return Number(BigNumber(value).toFixed(decimals)) // TODO only use bignumber
    },
  },
  validations() {
    return {
      address: {
        required,
        validAddress: (address) =>
          this.bech32Validate(address) ||
          isPolkadotAddress(address) ||
          isStarname(address),
      },
      amount: {
        required: (x) => !!x && x !== `0`,
        decimal,
        max: (x) => Number(x) <= this.maxAmount,
        min: (x) => Number(x) >= SMALLEST,
        maxDecimals: (x) => {
          return x.toString().split(".").length > 1
            ? x.toString().split(".")[1].length <= 6
            : true
        },
      },
      denoms: { required },
      selectedToken: { required },
      memo: {
        maxLength: maxLength(this.max_memo_characters),
      },
    }
  },
  apollo: {
    balances: {
      query: gql`
        query BalancesSendModal($networkId: String!, $address: String!) {
          balances(networkId: $networkId, address: $address) {
            id
            amount
            denom
          }
        }
      `,
      /* istanbul ignore next */
      skip() {
        return (
          !this.userAddress ||
          !this.$refs.actionModal ||
          !this.$refs.actionModal.show
        )
      },
      /* istanbul ignore next */
      variables() {
        return {
          networkId: this.network,
          address: this.userAddress,
        }
      },
      update(data) {
        // if no token preselected, pick first token available
        if (!this.selectedToken) {
          data.balances.find(({ denom }) => denom).denom
        }
        return data.balances || []
      },
    },
    networkFees: {
      query: gql`
        query NetworkFees(
          $networkId: String!
          $messageType: String!
          $message: TransactionDetailsInput!
          $senderAddress: String!
        ) {
          networkFees(
            networkId: $networkId
            messageType: $messageType
            message: $message
            senderAddress: $senderAddress
          ) {
            transactionFee {
              denom
              amount
            }
          }
        }
      `,
      /* istanbul ignore next */
      variables() {
        let { type, ...message } = this.transactionData
        delete message.memo
        // make sure the amounts are strings when sending
        if (message.amount) {
          message.amount = {
            amount: String(message.amount.amount),
            denom: message.amount.denom,
          }
        }
        if (message.amounts) {
          message.amounts = message.amounts.map(({ amount, denom }) => ({
            amount: String(amount),
            denom,
          }))
        }
        return {
          networkId: this.network,
          messageType: type,
          message,
          senderAddress: this.userAddress,
        }
      },
      /* istanbul ignore next */
      update(data) {
        if (data.networkFees) {
          this.networkFeesLoaded = true
          return data.networkFees
        }
      },
      /* istanbul ignore next */
      skip() {
        return (
          !this.userAddress ||
          !this.$refs.actionModal ||
          !this.$refs.actionModal.show ||
          !this.transactionData ||
          Object.keys(this.transactionData).length === 0
        )
      },
    },
    $subscribe: {
      userTransactionAdded: {
        /* istanbul ignore next */
        variables() {
          return {
            networkId: this.network,
            address: this.userAddress,
          }
        },
        /* istanbul ignore next */
        skip() {
          return (
            !this.userAddress ||
            !this.$refs.actionModal ||
            !this.$refs.actionModal.show
          )
        },
        query: UserTransactionAdded,
        /* istanbul ignore next */
        result() {
          this.$apollo.queries.balances.refetch()
        },
      },
    },
  },
}
</script>
<style scoped>
.tm-field-addon {
  border-right: 0;
}

.tm-field-addon:focus {
  border-color: var(--input-bc);
}

.tm-field-token-selector {
  width: 80px;
}

.tm-field-token-selector >>> .tm-field-select {
  border-left: 0;
  border-radius: 0 !important;
}

.tm-field-token-selector >>> .tm-field-select:focus {
  border-color: var(--input-bc);
}

.tm-field-token-selector >>> .tm-field-select-addon {
  border: 0;
}

.memo-span {
  font-size: var(--sm);
  font-style: italic;
}
</style>
