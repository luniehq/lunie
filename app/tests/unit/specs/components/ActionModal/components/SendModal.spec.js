import { shallowMount, createLocalVue } from "@vue/test-utils"
import Vuelidate from "vuelidate"
import SendModal from "src/ActionModal/components/SendModal"
import AsyncComputed from "vue-async-computed"

describe(`SendModal`, () => {
  const localVue = createLocalVue()
  localVue.use(Vuelidate)
  localVue.use(AsyncComputed)
  localVue.directive(`focus`, () => {})

  let wrapper, $store

  const getters = {
    connected: true,
    address: "cosmos1thyn8gfapk2d0zsp6dysn99ynhcs2y759kwznx",
    network: "cosmos-hub-mainnet",
    networks: [
      {
        id: "cosmos-hub-mainnet",
        coinLookup: [
          { viewDenom: "STAKE", chainToViewConversionFactor: 0.000001 },
        ],
      },
      {
        id: "terra-mainnet",
        coinLookup: [
          { viewDenom: "LUNA", chainToViewConversionFactor: 0.000001 },
        ],
      },
    ],
    stakingDenom: "STAKE",
  }

  const state = {}

  beforeEach(async () => {
    $store = {
      commit: jest.fn(),
      dispatch: jest.fn(),
      getters,
      state,
    }

    wrapper = shallowMount(SendModal, {
      localVue,
      mocks: {
        $store,
        $apollo: {
          queries: {
            balances: {
              refetch: () => {},
            },
          },
        },
      },
      propsData: {
        denoms: ["STAKE"],
      },
      sync: false,
    })

    wrapper.setData({
      balances: [
        {
          denom: `STAKE`,
          amount: 10000,
        },
      ],
      selectedToken: "STAKE",
    })

    wrapper.vm.$refs.actionModal = {
      submit: (cb) => cb(),
      open: jest.fn(),
    }
    wrapper.vm.open("STAKE")
  })

  it(`should display send modal form`, async () => {
    await waitForTxDataLoaded(wrapper)
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`clears on close`, () => {
    const self = {
      $v: { $reset: jest.fn() },
      address: `cosmos1thyn8gfapk2d0zsp6dysn99ynhcs2y759kwznx`,
      amount: 10,
    }
    SendModal.methods.clear.call(self)
    expect(self.$v.$reset).toHaveBeenCalled()
    expect(self.address).toBe(undefined)
    expect(self.amount).toBe(undefined)
  })

  describe(`validation`, () => {
    it(`should show address required error`, async () => {
      wrapper.setProps({
        denom: `STAKE`,
      })
      wrapper.setData({
        address: ``,
        amount: 2,
      })
      wrapper.vm.validateForm()
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.$v.$error).toBe(true)
      await waitForTxDataLoaded(wrapper)
      expect(wrapper.element).toMatchSnapshot()
    })

    it(`should show bech32 error when address is not bech32`, async () => {
      wrapper.setProps({
        denom: `STAKE`,
      })
      wrapper.setData({
        address: `cosmos1thyn8gfapk2d0zsp6dysn99ynhcs2y759kwznx1234767`,
        amount: 2,
      })
      const valid = wrapper.vm.validateForm()
      expect(valid).toBe(false)
      await wrapper.vm.$nextTick()
      await waitForTxDataLoaded(wrapper)
      expect(wrapper.element).toMatchSnapshot()
    })

    it(`should show an error if trying to send more then the max amount of tokens`, async () => {
      wrapper.setProps({
        denom: `STAKE`,
      })
      wrapper.setData({
        address: `cosmos1thyn8gfapk2d0zsp6dysn99ynhcs2y759kwznx1234767`,
        amount: 1000,
      })
      const valid = wrapper.vm.validateForm()
      expect(valid).toBe(false)
      await wrapper.vm.$nextTick()
      await waitForTxDataLoaded(wrapper)
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  it(`should submit when enterPressed is called`, async () => {
    const self = {
      $refs: { actionModal: { validateChangeStep: jest.fn() } },
    }
    SendModal.methods.enterPressed.call(self)
    expect(self.$refs.actionModal.validateChangeStep).toHaveBeenCalled()
  })

  it(`should set selectedToken to the first balance denom in balances`, () => {
    wrapper.setData({
      balances: [
        {
          denom: `STAKE`,
          amount: 10000,
        },
      ],
      selectedToken: undefined,
    })
    wrapper.vm.open()
    expect(wrapper.vm.selectedToken).toBe("STAKE")
  })

  it(`should refocus on amount when focusOnAmount is called`, async () => {
    const self = {
      $refs: { amount: { $el: { focus: jest.fn() } } },
    }
    SendModal.methods.refocusOnAmount.call(self)
    expect(self.$refs.amount.$el.focus).toHaveBeenCalled()
  })

  it(`validates bech32 addresses`, () => {
    expect(
      wrapper.vm.bech32Validate(`cosmos1x7wzdumfj8pncd99mqc0mkqfrrps3l3pjz8tk6`)
    ).toBe(true)
    expect(
      wrapper.vm.bech32Validate(`cosmos15ky9du8a2wlstz6fpx3p4mqpjyrm5ctpesxxn9`)
    ).toBe(false)
  })

  it("should return transaction data in correct form", async () => {
    wrapper.setProps({
      denom: `STAKE`,
    })
    wrapper.setData({
      address: `cosmos12345`,
      amount: 2,
    })
    await waitForTxDataLoaded(wrapper)
    expect(wrapper.vm.transactionData).toEqual({
      type: "SendTx",
      amount: {
        amount: 2,
        denom: "STAKE",
      },
      memo: "",
      to: ["cosmos12345"],
      from: ["cosmos1thyn8gfapk2d0zsp6dysn99ynhcs2y759kwznx"],
    })
  })

  it("should return empty transaction data if amount is NaN", () => {
    wrapper.setProps({
      denom: `STAKE`,
    })
    wrapper.setData({
      address: `cosmos12345`,
      amount: `NaN`,
    })
    expect(wrapper.vm.transactionData).toEqual({})
  })

  it(`sends an event on success`, () => {
    const self = {
      $emit: jest.fn(),
    }
    SendModal.methods.onSuccess.call(self)
    expect(self.$emit).toHaveBeenCalledWith(
      "success",
      expect.objectContaining({})
    )
  })

  it("should return notification message", () => {
    wrapper.setProps({
      denom: `STAKE`,
    })
    wrapper.setData({
      address: `cosmos12345`,
      amount: 2,
    })
    expect(wrapper.vm.notifyMessage).toEqual({
      title: `Successful Send`,
      body: `Successfully sent 2 STAKEs to cosmâ€¦2345`,
    })
  })

  describe(`if amount field max button clicked`, () => {
    it(`amount has to be 10000 atom`, async () => {
      wrapper.setProps({
        denom: `STAKE`,
      })
      wrapper.setData({
        address: `cosmos12345`,
      })
      wrapper.vm.setMaxAmount()
      expect(wrapper.vm.amount).toBe(10000)
    })
    it(`should show warning message`, async () => {
      wrapper.setProps({
        denom: `STAKE`,
      })
      wrapper.setData({
        address: `cosmos12345`,
      })
      wrapper.vm.setMaxAmount()
      await wrapper.vm.$nextTick()
      expect(wrapper.html()).toContain(
        "You are about to use all your tokens for this transaction. Consider leaving a little bit left over to cover the network fees."
      )
    })
    it(`should not show warning message if balance = 0`, async () => {
      wrapper.setData({
        balances: [
          {
            amount: 0,
            denom: "STAKE",
          },
        ],
      })
      wrapper.vm.setMaxAmount()
      await wrapper.vm.$nextTick()
      expect(wrapper.html()).not.toContain(
        "You are about to use all your tokens for this transaction. Consider leaving a little bit left over to cover the network fees."
      )
    })
    it(`isMaxAmount() should return false if balance = 0`, async () => {
      wrapper.setData({
        balances: [
          {
            amount: 0,
            denom: "STAKE",
          },
        ],
      })
      wrapper.vm.setMaxAmount()
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.isMaxAmount()).toBe(false)
    })

    it(`setMaxAmount should deduct the Terra tax from total balance when sending alt-tokens in Terra`, () => {
      const self = {
        network: "terra-mainnet",
        amount: 0.000001,
        selectedBalance: {
          amount: 1,
          denom: "STAKE",
        },
        getTerraTax: SendModal.methods.getTerraTax,
        maxDecimals: SendModal.methods.maxDecimals,
        networkFeesLoaded: true,
        networkFees: {
          transactionFee: {
            amount: 0.0075,
            denom: "STAKE",
          },
        },
      }
      const maxAmount = SendModal.computed.maxAmount.call(self)
      expect(maxAmount).toBe(0.9925)
      self.maxAmount = maxAmount
      SendModal.methods.setMaxAmount.call(self)
      expect(self.amount).toBe(0.9925)
    })
  })

  describe(`Set token and balance`, () => {
    it(`it takes the corresponding balance from the balances array when
    selectedToken has been chosen and balances has a length over 1`, async () => {
      wrapper.setData({
        selectedToken: `TOKEN1`,
        balances: [
          {
            amount: 1,
            denom: "TOKEN1",
          },
          {
            amount: 2,
            denom: "TOKEN2",
          },
        ],
      })
      expect(wrapper.vm.selectedBalance.amount).toBe(1)
    })

    it(`it automatically sets the token to the first token in the balances`, async () => {
      wrapper.setData({
        selectedToken: ``,
        balances: [
          {
            amount: 1,
            denom: "STAKE",
          },
        ],
      })
      SendModal.watch.balances.call(wrapper.vm, [
        {
          amount: 1,
          denom: "STAKE",
        },
      ])
      expect(wrapper.vm.selectedBalance.amount).toBe(1)
    })
    // This one creates a lot of ugly errors
    // it(`returns empty string if selectedToken hasn't been chosen yet`, () => {
    //   wrapper.setData({
    //     balances: []
    //   })
    //   const res = wrapper.vm.token()
    //   expect(res).toBe("")
    // })
    it(`returns selectedToken if selectedToken has been chosen`, () => {
      const res = wrapper.vm.token()
      expect(res).toBe("STAKE")
    })
  })
})

async function waitForTxDataLoaded(wrapper) {
  // need to wait for transactionData to be resolved (very dirty solution)
  while (wrapper.vm.$asyncComputed.transactionData.updating) {
    await new Promise((resolve) => setTimeout(resolve, 100))
  }
  await new Promise((resolve) => setTimeout(resolve, 100))
}
