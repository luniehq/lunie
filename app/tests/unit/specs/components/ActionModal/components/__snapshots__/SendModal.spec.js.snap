// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendModal should display send modal form 1`] = `
<actionmodal-stub
  featureflag="send"
  id="send-modal"
  notifymessage="[object Object]"
  rewards=""
  selecteddenom="STAKE"
  submissionerrorprefix="Sending tokens failed"
  title="Send"
  transactiondata="[object Object]"
  transactiontype="SendTx"
  validate="function () { [native code] }"
>
  <tmformgroup-stub
    class="action-modal-form-group"
    fieldid="send-address"
    fieldlabel="Send To"
  >
    <tmfield-stub
      id="send-address"
      placeholder="Address or Starname"
      type="text"
      value=""
    />
     
    <!---->
     
    <span
      class="memo-span"
    >
      
      Lunie supports
      
      <a
        href="https://starname.me/"
        target="_blank"
      >
        IOV Starnames
      </a>
    </span>
  </tmformgroup-stub>
   
  <tmformgroup-stub
    class="action-modal-form-group"
    fieldid="amount"
    fieldlabel="Amount"
    id="form-group-amount"
  >
    <tmfieldgroup-stub>
      <tmfield-stub
        class="tm-field-addon"
        id="amount"
        placeholder="0"
        type="number"
      />
       
      <tmfield-stub
        class="tm-field-token-selector"
        id="token"
        options="[object Object]"
        placeholder="Select the token"
        title="Select the token you wish to operate with"
        type="select"
        value="STAKE"
      />
       
      <tmbtn-stub
        class="addon-max"
        customclass=""
        type="button"
        value="Set Max"
      />
    </tmfieldgroup-stub>
     
    <!---->
  </tmformgroup-stub>
   
  <tmformgroup-stub
    class="action-modal-group"
    fieldid="memo"
    fieldlabel="Memo"
    id="memo"
  >
    <tmfield-stub
      id="memo"
      type="text"
      value=""
    />
     
    <span
      class="memo-span"
    >
      To learn more about how to use the memo field, read
      
      <a
        href=" https://intercom.help/lunie/en/articles/3776563-using-the-memo-option-when-sending-tokens-to-and-from-exchanges"
        rel="noopener norefferer"
        target="_blank"
      >
        our guide
      </a>
      .
    </span>
     
    <!---->
  </tmformgroup-stub>
</actionmodal-stub>
`;

exports[`SendModal validation should show address required error 1`] = `
<actionmodal-stub
  amount="2"
  featureflag="send"
  id="send-modal"
  notifymessage="[object Object]"
  rewards=""
  selecteddenom="STAKE"
  submissionerrorprefix="Sending tokens failed"
  title="Send"
  transactiondata="[object Object]"
  transactiontype="SendTx"
  validate="function () { [native code] }"
>
  <tmformgroup-stub
    class="action-modal-form-group"
    error="true"
    fieldid="send-address"
    fieldlabel="Send To"
  >
    <tmfield-stub
      id="send-address"
      placeholder="Address or Starname"
      type="text"
      value=""
    />
     
    <tmformmsg-stub
      msg=""
      name="Address"
      type="required"
    />
     
    <span
      class="memo-span"
    >
      
      Lunie supports
      
      <a
        href="https://starname.me/"
        target="_blank"
      >
        IOV Starnames
      </a>
    </span>
  </tmformgroup-stub>
   
  <tmformgroup-stub
    class="action-modal-form-group"
    fieldid="amount"
    fieldlabel="Amount"
    id="form-group-amount"
  >
    <tmfieldgroup-stub>
      <tmfield-stub
        class="tm-field-addon"
        id="amount"
        placeholder="0"
        type="number"
        value="2"
      />
       
      <tmfield-stub
        class="tm-field-token-selector"
        id="token"
        options="[object Object]"
        placeholder="Select the token"
        title="Select the token you wish to operate with"
        type="select"
        value="STAKE"
      />
       
      <tmbtn-stub
        class="addon-max"
        customclass=""
        type="button"
        value="Set Max"
      />
    </tmfieldgroup-stub>
     
    <!---->
  </tmformgroup-stub>
   
  <tmformgroup-stub
    class="action-modal-group"
    fieldid="memo"
    fieldlabel="Memo"
    id="memo"
  >
    <tmfield-stub
      id="memo"
      type="text"
      value=""
    />
     
    <span
      class="memo-span"
    >
      To learn more about how to use the memo field, read
      
      <a
        href=" https://intercom.help/lunie/en/articles/3776563-using-the-memo-option-when-sending-tokens-to-and-from-exchanges"
        rel="noopener norefferer"
        target="_blank"
      >
        our guide
      </a>
      .
    </span>
     
    <!---->
  </tmformgroup-stub>
</actionmodal-stub>
`;

exports[`SendModal validation should show an error if trying to send more then the max amount of tokens 1`] = `
<actionmodal-stub
  amount="1000"
  featureflag="send"
  id="send-modal"
  notifymessage="[object Object]"
  rewards=""
  selecteddenom="STAKE"
  submissionerrorprefix="Sending tokens failed"
  title="Send"
  transactiondata="[object Object]"
  transactiontype="SendTx"
  validate="function () { [native code] }"
>
  <tmformgroup-stub
    class="action-modal-form-group"
    error="true"
    fieldid="send-address"
    fieldlabel="Send To"
  >
    <tmfield-stub
      id="send-address"
      placeholder="Address or Starname"
      type="text"
      value="cosmos1thyn8gfapk2d0zsp6dysn99ynhcs2y759kwznx1234767"
    />
     
    <tmformmsg-stub
      msg="doesn't have a format known by Lunie"
      name="Address"
      type="custom"
    />
     
    <span
      class="memo-span"
    >
      
      Lunie supports
      
      <a
        href="https://starname.me/"
        target="_blank"
      >
        IOV Starnames
      </a>
    </span>
  </tmformgroup-stub>
   
  <tmformgroup-stub
    class="action-modal-form-group"
    fieldid="amount"
    fieldlabel="Amount"
    id="form-group-amount"
  >
    <tmfieldgroup-stub>
      <tmfield-stub
        class="tm-field-addon"
        id="amount"
        placeholder="0"
        type="number"
        value="1000"
      />
       
      <tmfield-stub
        class="tm-field-token-selector"
        id="token"
        options="[object Object]"
        placeholder="Select the token"
        title="Select the token you wish to operate with"
        type="select"
        value="STAKE"
      />
       
      <tmbtn-stub
        class="addon-max"
        customclass=""
        type="button"
        value="Set Max"
      />
    </tmfieldgroup-stub>
     
    <!---->
  </tmformgroup-stub>
   
  <tmformgroup-stub
    class="action-modal-group"
    fieldid="memo"
    fieldlabel="Memo"
    id="memo"
  >
    <tmfield-stub
      id="memo"
      type="text"
      value=""
    />
     
    <span
      class="memo-span"
    >
      To learn more about how to use the memo field, read
      
      <a
        href=" https://intercom.help/lunie/en/articles/3776563-using-the-memo-option-when-sending-tokens-to-and-from-exchanges"
        rel="noopener norefferer"
        target="_blank"
      >
        our guide
      </a>
      .
    </span>
     
    <!---->
  </tmformgroup-stub>
</actionmodal-stub>
`;

exports[`SendModal validation should show bech32 error when address is not bech32 1`] = `
<actionmodal-stub
  amount="2"
  featureflag="send"
  id="send-modal"
  notifymessage="[object Object]"
  rewards=""
  selecteddenom="STAKE"
  submissionerrorprefix="Sending tokens failed"
  title="Send"
  transactiondata="[object Object]"
  transactiontype="SendTx"
  validate="function () { [native code] }"
>
  <tmformgroup-stub
    class="action-modal-form-group"
    error="true"
    fieldid="send-address"
    fieldlabel="Send To"
  >
    <tmfield-stub
      id="send-address"
      placeholder="Address or Starname"
      type="text"
      value="cosmos1thyn8gfapk2d0zsp6dysn99ynhcs2y759kwznx1234767"
    />
     
    <tmformmsg-stub
      msg="doesn't have a format known by Lunie"
      name="Address"
      type="custom"
    />
     
    <span
      class="memo-span"
    >
      
      Lunie supports
      
      <a
        href="https://starname.me/"
        target="_blank"
      >
        IOV Starnames
      </a>
    </span>
  </tmformgroup-stub>
   
  <tmformgroup-stub
    class="action-modal-form-group"
    fieldid="amount"
    fieldlabel="Amount"
    id="form-group-amount"
  >
    <tmfieldgroup-stub>
      <tmfield-stub
        class="tm-field-addon"
        id="amount"
        placeholder="0"
        type="number"
        value="2"
      />
       
      <tmfield-stub
        class="tm-field-token-selector"
        id="token"
        options="[object Object]"
        placeholder="Select the token"
        title="Select the token you wish to operate with"
        type="select"
        value="STAKE"
      />
       
      <tmbtn-stub
        class="addon-max"
        customclass=""
        type="button"
        value="Set Max"
      />
    </tmfieldgroup-stub>
     
    <!---->
  </tmformgroup-stub>
   
  <tmformgroup-stub
    class="action-modal-group"
    fieldid="memo"
    fieldlabel="Memo"
    id="memo"
  >
    <tmfield-stub
      id="memo"
      type="text"
      value=""
    />
     
    <span
      class="memo-span"
    >
      To learn more about how to use the memo field, read
      
      <a
        href=" https://intercom.help/lunie/en/articles/3776563-using-the-memo-option-when-sending-tokens-to-and-from-exchanges"
        rel="noopener norefferer"
        target="_blank"
      >
        our guide
      </a>
      .
    </span>
     
    <!---->
  </tmformgroup-stub>
</actionmodal-stub>
`;
